brk.c

```diff
@@ -47,46 +47,49 @@
 #define HEAP_LEN   PAGES2BYTES(CONFIG_APPELFLOADER_BRK_NBPAGES)
 
 /*
- * For now we only support one custom stack
+ * For now we only support one custom heap
  */
 static void *base = NULL;
+static void *brk_cur = NULL;
 static void *zeroed = NULL;
 static intptr_t len = 0;
 
 UK_LLSYSCALL_R_DEFINE(void *, brk, void *, addr)
 {
-	if (!addr && base) {
-		/* another brk context request although we have one already */
-		uk_pr_crit("Cannot handle multiple user space heaps: Not implemented!\n");
-		return ERR2PTR(-ENOMEM);
-	}
-
 	if (!addr) {
 		/* allocate brk context */
-		base = uk_palloc(uk_alloc_get_default(), HEAP_PAGES);
 		if (!base) {
-			uk_pr_crit("Could not allocate memory for heap (%"PRIu64" KiB): Out of memory\n",
-				   (uint64_t) HEAP_LEN / 1024);
-			return ERR2PTR(-ENOMEM);
+			base = uk_palloc(uk_alloc_get_default(), HEAP_PAGES);
+			if (!base) {
+				uk_pr_crit("Could not allocate memory for heap (%"PRIu64" KiB): Out of memory\n",
+					(uint64_t) HEAP_LEN / 1024);
+				return ERR2PTR(-ENOMEM);
+			}
+
+			/* initialize brk_cur with start of allocated heap region */
+			brk_cur = base;
 		}
 		zeroed = base;
```

Comment:
I think this statement should be moved up, below Line 70. We should reset `zeroed` only to base if the heap region got allocated.